### DISCUSSION TITLE
Using a figure layout with computational images with different dimensions for the subfigures

### DISCUSSION THREAD

**Discussion by andrewheiss** [quarto-dev/quarto-cli#7022] [Q&A] [answered]
### Description

Currently it's possible to use a code chunk to generate a figure layout with the images created by the chunk, like so:

````qmd
See @fig-example-same-size-1 and @fig-example-same-size-2

```{r}
#| echo: false
#| label: fig-example-same-size
#| fig-cap: "Charts"
#| fig-subcap: 
#|   - "Cars"
#|   - "Pressure"
#| layout-ncol: 2

plot(cars)
plot(pressure)
```
````

<img width="820" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/73663/367ec0bf-e69f-4485-8633-8f6d8235dbbc">

&nbsp; 

It's possible to control the widths of those subplots too with the `layout` option:

````qmd
See @fig-example-different-sizes-1 and @fig-example-different-sizes-2

```{r}
#| echo: false
#| label: fig-example-different-sizes
#| fig-cap: "Charts"
#| fig-subcap: 
#|   - "Cars"
#|   - "Pressure"
#| layout: "[[ 25, 75 ]]"

plot(cars)
plot(pressure)
```
````

<img width="834" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/73663/5e23cfa6-bdc6-408b-9825-1f85ea21c004">

&nbsp;

However, the dimensions are all squished because both images generated by the chunk use the same width and height.

It would be neat to be able to feed a list of dimensions to `fig.width` and/or `fig.height` so that each of the generated images use different dimensions, but I don't think that's possible (see this issue over at {knitr} https://github.com/yihui/knitr/issues/1517)

As an alternative, it's possible to put the different figures in separate chunks and use a `:::` container to build the layout:

````qmd

See @fig-example-multiple-chunks-1 and @fig-example-multiple-chunks-2.

::: {#fig-example-multiple-chunks layout="[[ 25, 75 ]]"}

```{r fig1, echo=FALSE, fig.width=2, fig.height=5}
#| fig-subcap: "Cars"
plot(cars)
```

```{r fig2, echo=FALSE, fig.width=6, fig.height=5}
#| fig-subcap: "Pressure"
plot(pressure)
```

Charts

:::
````

This looks great! 

<img width="803" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/73663/cd16f663-cd9c-4a78-8624-93577d6bcc18">

&nbsp;

But it no longer treats the plots as subfigures. The cross referencing is broken ("See ?@fig-example-multiple-chunks-1" appears in the text), and the captions don't have the "(a)" and "(b)" prefixes.

Is there a way to get differently-sized computational images into a custom figure layout like this? Either with a vector of dimensions (like `fig.width = c(2, 6), fig.height = 5` or by somehow building it with a `:::` block?

**Comment by mcanouil** [comment]
This should work `fig.width = c(2, 6), fig.height = 5`, as the two parameters are vectorised, or they were at some point.

Note that you should not mix "inline R option" and "yaml option" in the same code cell or expect unintended behaviour.

**Reply by andrewheiss** [reply]
That creates two plots, but they still have the same dimensions:

````
```{r fig.width = c(2, 6), fig.height = 5}
#| echo: false
#| label: fig-example-different-sizes-1
#| fig-cap: "Charts"
#| fig-subcap: 
#|   - "Cars"
#|   - "Pressure"
#| layout: "[[ 25, 75 ]]"

plot(cars)
plot(pressure)
```
````

<img width="832" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/73663/2b5b83ec-5f24-4957-8fdd-35f341667033">

&nbsp;

And doing it with inline R makes Quarto mad, since `fig.height` and `fig.width` are hardcoded to only accept single numbers:

<img width="440" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/73663/d888f9d5-97b5-48c0-8a45-757ae098309e">

**Reply by andrewheiss** [reply]
`fig.width` and `fig.height` are vectorized, but only for different output devices and not for chunks that create multiple plots—this issue at {knitr} raised the possibility of adding support for multiple plots https://github.com/yihui/knitr/issues/1517

**Reply by mcanouil** [reply]
In yaml style, you need `#| fig-width: !expr c(2, 6)`.

But if `knitr` does not support this, there is not  much that can be done in Quarto itself.

A possibility would be to write a hook to deal with vectorise dimensions, but that's more of a hack.

**Reply by cscheid** [reply]
> In yaml style, you need `#| fig-width: !expr c(2, 6)`

No, please don't 😭  That will just confuse us even more. If the schema isn't accepting it, that's because our code probably won't work. If the code does work, then we need to change the schema!

**Reply by andrewheiss** [reply]
Yeah, the code doesn't work and knitr doesn't support it

Is there a way to get subfigures by making the layout manually like this? 

````
See @fig-example-multiple-chunks-1 and @fig-example-multiple-chunks-2.

::: {#fig-example-multiple-chunks layout="[[ 25, 75 ]]"}

```{r fig1, echo=FALSE, fig.width=2, fig.height=5}
#| fig-subcap: "Cars"
plot(cars)
```

```{r fig2, echo=FALSE, fig.width=6, fig.height=5}
#| fig-subcap: "Pressure"
plot(pressure)
```

Charts

:::
````

That works just fine with static images, like the elephants in the documentation:

<img width="802" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/73663/c9e397e1-b1b0-4c76-84b6-724fc7de4655">

but it doesn't work with computational output

**Reply by cscheid** [reply]
Like this:

````
---
title: Test
---

See @fig-subfig-1 and @fig-subfig-2.

::: {#fig-example-multiple-chunks layout="[[ 25, 75 ]]"}

::: {#fig-subfig-1}
```{r}
#| echo: false
#| fig-width: 2
#| fig-height: 5
plot(cars)
```

Cars

:::

::: {#fig-subfig-2}
```{r}
#| echo: false
#| fig-width: 6
#| fig-height: 5
plot(pressure)
```

Pressure

:::

Charts

:::
````

**Reply by cscheid** [reply]
NB: We should make something like your previous syntax work. But in Quarto 1.4 prereleases, we're a lot more lax about what constitutes figures and subfigures, and so you can get away with stuff like what I just wrote.

(Note also that if you `echo: true`, then you might end up with code where you don't expect it to! So this isn't actually perfect. But it should be enough to get you going.)

**Reply by andrewheiss** [reply]
Ahhh that works! It's definitely more verbose than the static elephant syntax, but it works.

<img width="804" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/73663/ac4bc4d0-d3a3-43a7-b05a-73a704bedd29">

````qmd
See @fig-subfig-1 and @fig-subfig-2.

:::: {#fig-example-multiple-chunks layout="[[ 25, 75 ]]"}

::: {#fig-subfig-1}
```{r}
#| echo: false
#| fig-width: 2
#| fig-height: 5
plot(cars)
```

Cars

:::

::: {#fig-subfig-2}
```{r}
#| echo: false
#| fig-width: 6
#| fig-height: 5
plot(pressure)
```

Pressure

:::

Charts

::::

````

**Reply by cscheid** [reply, marked as answer]
Btw, this also works (surprisingly to me!)

````
---
title: Test
---

See @fig-subfig-1 and @fig-subfig-2.

::: {#fig-example-multiple-chunks layout="[[ 25, 75 ]]"}

```{r}
#| label: fig-subfig-1
#| fig-cap: Cars
#| echo: false
#| fig-width: 2
#| fig-height: 5
plot(cars)
```

```{r}
#| label: fig-subfig-2
#| fig-cap: Pressure
#| echo: false
#| fig-width: 6
#| fig-height: 5
plot(pressure)
```

Charts

:::
````

**Reply by andrewheiss** [reply]
That's even better, and it's actually basically what I had in the original question, where I ended up with this output with the wrong crossreferences: 

![image](https://github.com/quarto-dev/quarto-cli/assets/73663/d60c0aca-132c-4cff-b067-e782e15efb01)

&nbsp;

But that's because I was thinking it was like the output from a multi-plot chunk, so I was using `fig-example-multiple-chunks-1` and `fig-example-multiple-chunks-2` to reference items 1 and 2 inside the figure div, which didn't work. Also, I wasn't using `fig-X` for those interior chunks, which was also breaking stuff. 

Referencing the labels inside directly works, though. 

So in my original question, adding `fig-` to the interior chunks and then referencing them fixes everything:

````qmd
See @fig-fig1 and @fig-fig2.

::: {#fig-example-multiple-chunks layout="[[ 25, 75 ]]"}

```{r fig-fig1, echo=FALSE, fig.width=2, fig.height=5}
#| fig-subcap: "Cars"
plot(cars)
```

```{r fig-fig2, echo=FALSE, fig.width=6, fig.height=5}
#| fig-subcap: "Pressure"
plot(pressure)
```

Charts

:::
````

**Reply by cscheid** [reply]
The general principle is that "anything goes" as the content of a layout block. But if the inner content has figure-like crossref information, then we should be able to parse it as a subfigure and make the layout work.

Hopefully that will help clean up a lot of previously-necessary-but-ugly quarto code out there. (and sorry again about the regressions..)

**Reply by cscheid** [reply]
@andrewheiss Just please please, can we convince you to get in the habit of using YAML instead of knitr chunk options? We can't autocomplete or validate knitr chunk options, so those are going to be even riskier going forward...

**Reply by andrewheiss** [reply]
Oh, is there a practical difference between using `fig-subcap` and `fig-cap` here? In your example you used `fig-cap` for the interior plots; I used `fig-subcap`. The output seems to be the same. Is that a fluke, or part of the "anything goes" approach?

**Reply by cscheid** [reply]
> But that's because I was thinking it was like the output from a multi-plot chunk

To be clear, quarto _should_ work like that, but it doesn't right now. We'll fix it in the future, but now that the general crossref infra is a bit more robust, this will probably be lower priority, since the general figure structure works for subfigs too.

**Reply by cscheid** [reply]
> Oh, is there a practical difference between using fig-subcap and fig-cap here? In your example you used fig-cap for the interior plots; I used fig-subcap. The output seems to be the same. Is that a fluke, or part of the "anything goes" approach?

Let's call it a happy coincidence! The truth is I was myself surprised that this works. I still need to understand _why_ so we don't regress.

**Reply by mcanouil** [reply]
And if you want the code to echo there is still a way via `knitr`:

<img width="727" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/a8fcffad-5a3a-438f-a860-3bf5f4a710d3">


````qmd
---
title: Test
format: html
---

See @fig-subfig-1 and @fig-subfig-2.

::: {#fig-example-multiple-chunks layout="[[ 25, 75 ]]"}

```{r}
#| label: subfig-1
#| echo: true
#| eval: false
plot(cars)
```

```{r}
#| label: subfig-2
#| echo: true
#| eval: false
plot(pressure)
```

```{r}
#| label: fig-subfig-1
#| ref-label: subfig-1
#| eval: true
#| echo: false
#| fig-width: 2
#| fig-height: 5
```

```{r}
#| label: fig-subfig-2
#| ref-label: subfig-2
#| eval: true
#| echo: false
#| fig-width: 6
#| fig-height: 5
```

Charts

:::

````

**Reply by andrewheiss** [reply]
Oh awesome!

**Reply by mcanouil** [reply]
Actually, just `echo: true` do the job without messing up the layout.

> (Note also that if you echo: true, then you might end up with code where you don't expect it to! So this isn't actually perfect. But it should be enough to get you going.)

@cscheid seems pretty much "perfect" 😉

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/65a03fa4-4390-4f79-ad2c-4becdc0e1499">


````qmd
---
title: Test
format: html
---

See @fig-subfig-1 and @fig-subfig-2.

::: {#fig-example-multiple-chunks layout="[[ 25, 75 ]]"}

```{r}
#| label: fig-subfig-1
#| echo: true
#| fig-width: 2
#| fig-height: 5
#| fig-cap: "This is a caption."
plot(cars)
```

```{r}
#| label: fig-subfig-2
#| echo: true
#| fig-width: 6
#| fig-height: 5
#| fig-cap: "This is another caption."
plot(pressure)
```

Charts

:::
````

**Comment by andrewheiss** [comment]
> @andrewheiss Just please please, can we convince you to get in the habit of using YAML instead of knitr chunk options? We can't autocomplete or validate knitr chunk options, so those are going to be even riskier going forward...

@cscheid Ha yes, force of habit. But also I've been keeping only fig.width and fig.height up in the chunk options because it seems to be the only way RStudio can recognize the correct dimensions (via https://github.com/rstudio/rstudio/issues/4521), so my normal approach has been to put all chunk options as YAML except for dimensions. But that's a whole separate issue :)

<img width="942" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/73663/18d138a9-2a3a-45d0-a51e-4a2acdfd30bb">

**Reply by andrewheiss** [reply]
Continued the issue over there: https://github.com/rstudio/rstudio/issues/4521#issuecomment-1736204638
